src/gluonts/mx/distribution/isqf.py

```diff
@@ -0,0 +1,966 @@
+    by Park, Robinson, Aubet, Kan, Gasthaus, Wang
+
+    Parameters
+    ----------
+    spline_knots, spline_heights
+        Tensor parametrizing the x-positions (y-positions) of the spline knots
+        Shape: (*batch_shape, (num_qk-1), num_pieces)
+    qk_x, qk_y
+        Tensor containing the increasing x-positions (y-positions) of the quantile knots
+        Shape: (*batch_shape, num_qk)
+    beta_l, beta_r
+        Tensor containing the non-negative learnable parameter of the left (right) tail
+        Shape: (*batch_shape,)
+    """
+
+    is_reparameterizable = False
+
+    @validated()
+    def __init__(
+        self,
+        spline_knots: Tensor,
+        spline_heights: Tensor,
+        beta_l: Tensor,
+        beta_r: Tensor,
+        qk_y: Tensor,
+        qk_x: Tensor,
+        num_qk: int,
+        num_pieces: int,
+        tol: float = 1e-4,
+    ) -> None:
+        self.num_qk, self.num_pieces = num_qk, num_pieces
+        self.spline_knots, self.spline_heights = spline_knots, spline_heights
+        self.beta_l, self.beta_r = beta_l, beta_r
+        self.qk_y_all = qk_y
+        self.tol = tol
+
+        F = self.F
+
+        # Get quantile knots (qk) parameters
+        (
+            self.qk_x,
+            self.qk_x_plus,
+            self.qk_x_l,
+            self.qk_x_r,
+        ) = ISQF.parametrize_qk(F, qk_x)
+        (
+            self.qk_y,
+            self.qk_y_plus,
+            self.qk_y_l,
+            self.qk_y_r,
+        ) = ISQF.parametrize_qk(F, qk_y)
+
+        # Get spline knots (sk) parameters
+        self.sk_y, self.delta_sk_y = ISQF.parametrize_spline(
+            F,
+            self.spline_heights,
+            self.qk_y,
+            self.qk_y_plus,
+            self.num_pieces,
+            self.tol,
+        )
+        self.sk_x, self.delta_sk_x = ISQF.parametrize_spline(
+            F,
+            self.spline_knots,
+            self.qk_x,
+            self.qk_x_plus,
+            self.num_pieces,
+            self.tol,
+        )
+
+        if self.num_pieces > 1:
+            self.sk_x_plus = F.concat(
+                F.slice_axis(self.sk_x, axis=-1, begin=1, end=None),
+                F.expand_dims(self.qk_x_plus, axis=-1),
+                dim=-1,
+            )
+        else:
+            self.sk_x_plus = F.expand_dims(self.qk_x_plus, axis=-1)
+
+        # Get tails parameters
+        self.tail_al, self.tail_bl = ISQF.parametrize_tail(
+            F, self.beta_l, self.qk_x_l, self.qk_y_l
+        )
+        self.tail_ar, self.tail_br = ISQF.parametrize_tail(
+            F, -self.beta_r, 1 - self.qk_x_r, self.qk_y_r
+        )
+
+    @property
+    def F(self):
+        return getF(self.beta_l)
+
+    @property
+    def args(self) -> List:
+        return [
+            self.spline_knots,
+            self.spline_heights,
+            self.beta_l,
+            self.beta_r,
+            self.qk_y_all,
+        ]
+
+    @staticmethod
```

Comment:
I'm not a fan of these staticmethods. Couldn't we just declare these as functions?